fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/ResourceReferenceDAO.java

```diff
@@ -756,78 +753,104 @@ public void upsertCommonTokenValues(List<ResourceTokenValueRec> values) {
-            paramList.append("(CAST(? AS VARCHAR(" + FhirSchemaConstants.MAX_TOKEN_VALUE_BYTES + "))");
-            paramList.append(",CAST(? AS INT))");
+        Optional<Integer> maxQuery = translator.maximumQueryParameters();
 
-            // also build the inList for the select statement later
-            if (inList.length() > 0) {
-                inList.append(",");
-            }
-            inList.append("(?,?)");
-        }
+        // Process segments up to maximum query parameter size.
+        for (int low = 0; low < sortedTokenValues.size();/*at end  of loop*/) {
+            int high = sortedTokenValues.size();
 
-        final String paramListStr = paramList.toString();
-        doCommonTokenValuesUpsert(paramListStr, sortedTokenValues);
+            if (!maxQuery.isEmpty() && high > maxQuery.get()/2) {
+                // 2 as there a pair inserted.
+                high = low + maxQuery.get()/2;
+            }
 
-        // Now grab the ids for the rows we just created. If we had a RETURNING implementation
-        // which worked reliably across all our database platforms, we wouldn't need this
-        // second query.
-        // Derby doesn't support IN LISTS with multiple members, so we have to join against
-        // a VALUES again. No big deal...probably similar amount of work for the database
-        StringBuilder select = new StringBuilder();
-        select.append("     SELECT ctv.code_system_id, ctv.token_value, ctv.common_token_value_id FROM ");
-        select.append("     (VALUES ").append(paramListStr).append(" ) AS v(token_value, code_system_id) ");
-        select.append("       JOIN common_token_values ctv ");
-        select.append("              ON ctv.token_value = v.token_value ");
-        select.append("             AND ctv.code_system_id = v.code_system_id ");
-
-        // Grab the ids
-        Map<CommonTokenValue, Long> idMap = new HashMap<>();
-        try (PreparedStatement ps = connection.prepareStatement(select.toString())) {
-            int a = 1;
-            for (CommonTokenValue tv: sortedTokenValues) {
-                ps.setString(a++, tv.getTokenValue());
-                ps.setInt(a++, tv.getCodeSystemId());
+            // Make sure we don't overflow.
+            if (high > sortedTokenValues.size()) {
+                high = sortedTokenValues.size();
             }
 
-            ResultSet rs = ps.executeQuery();
-            while (rs.next()) {
-                // SELECT code_system_id, token_value...note codeSystem not required
-                CommonTokenValue key = new CommonTokenValue(null, rs.getInt(1), rs.getString(2));
-                idMap.put(key, rs.getLong(3));
+            // The sublist as we have a maxQuery
+            List<CommonTokenValue> sortedTokenValuesSub;
+            if (!maxQuery.isEmpty()) {
+                sortedTokenValuesSub = sortedTokenValues.subList(low, high - 1);
+            } else {
+                sortedTokenValuesSub = sortedTokenValues;
```

Comment:
Up to here is new...

Reply:
What is the semantic impact of only using the subset of values? If we're hitting the limit here, I think we should be using an alternative strategy (perhaps a temporary table).

Reply:
It means that insert is at least 1 operation and potentially more.  `1..*` executions 

Semantically the `merge into` is probably the best option rather than a pseudo `upsert`. 

I'd say the likelihood of that many unique code systems values and ids is fairly low once a system is somewhat loaded with data, and the cache is active and warm.  Then the potential for multiple inserts is low.  

Yes, a temporary table is an option.  It's probably the better option in the longer run.

Reply:
I think we can come back to this when we revisit the cache stuff which will require us to move everything to transaction data and push everything to the database during commit. In some cases it might be easier just to use batch operations.